bundles/automation/org.eclipse.smarthome.automation.core/src/main/java/org/eclipse/smarthome/automation/core/RuleEngine.java

```diff
@@ -0,0 +1,426 @@
+                    result.add(new RuleImpl(r));
+                }
+            } else {
+                result.add(new RuleImpl(r));
+            }
+        }
+        return result;
+    }
+
+    public Collection<Rule> getRules(String filter) {
+        Collection<Rule> result = new ArrayList<Rule>(10);
+        for (Iterator<RuleImpl> it = rules.values().iterator(); it.hasNext();) {
+            RuleImpl r = it.next();
+            result.add(new RuleImpl(r));
+        }
+        return result;
+    }
+
+    protected Collection<String> getScopeIds() {
+        Set<String> result = new HashSet<String>(10);
+        for (Iterator<RuleImpl> it = rules.values().iterator(); it.hasNext();) {
+            RuleImpl r = it.next();
+            String id = r.getScopeIdentifier();
+            if (id != null) {
+                result.add(id);
+            }
+        }
+        return result;
+    }
+
+    /**
+     *
+     * @see org.osgi.util.tracker.ServiceTrackerCustomizer#addingService(org.osgi.framework.ServiceReference)
+     */
+    @Override
+    public ModuleHandlerFactory addingService(ServiceReference/* <ModuleHandlerFactory> */reference) {
+        ModuleHandlerFactory msf = (ModuleHandlerFactory) bc.getService(reference);
+        Collection<String> moduleTypes = msf.getTypes();
+        for (Iterator<String> it = moduleTypes.iterator(); it.hasNext();) {
+            String moduleTypeName = it.next();
+            if (moduleHandlerFactories.get(moduleTypeName) != null) {
+                // log.error
+            } else {
+                moduleHandlerFactories.put(moduleTypeName, msf);
+            }
+
+        }
+        return msf;
+    }
+
+    /**
+     *
+     * @see org.osgi.util.tracker.ServiceTrackerCustomizer#modifiedService(org.osgi.framework.ServiceReference,
+     *      java.lang.Object)
+     */
+    @Override
+    public void modifiedService(ServiceReference/* <ModuleHandlerFactory> */reference, /* ModuleHandlerFactory */
+            Object service) {
+        // TODO Auto-generated method stub
+
+    }
+
+    /**
+     * @see org.osgi.util.tracker.ServiceTrackerCustomizer#removedService(org.osgi.framework.ServiceReference,
+     *      java.lang.Object)
+     */
+    @Override
+    public void removedService(ServiceReference/* <ModuleHandlerFactory> */reference, /* ModuleHandlerFactory */
+            Object service) {
+        Collection<String> moduleTypes = ((ModuleHandlerFactory) service).getTypes();
+        for (Iterator<String> it = moduleTypes.iterator(); it.hasNext();) {
+            String moduleTypeName = it.next();
+            moduleHandlerFactories.remove(moduleTypeName);
+        }
+    }
+
+    public void runRule(RuleEngineCallbackImpl reCallback) {
+        synchronized (reCallback) {
+            Rule r = reCallback.getRule();
+            Thread t = runningRules.get(r.getUID());
+
+            if (t == null) {
+                RuleThread rt = new RuleThread(reCallback); // TODO use ThreadPool
```

Comment:
The trigger handled does not start the rule. It just add TriggerData to the queue of RuleEngineCallbackImpl. Also the thread which process TriggerData of the rule is reused in case there are other TriggerData added in the queue during the time of processing the last one.